server/implementation/src/main/java/io/smallrye/graphql/execution/event/EventEmitter.java

```diff
@@ -37,10 +37,11 @@
  */
 public class EventEmitter {
     private static final Logger LOG = Logger.getLogger(EventEmitter.class);
+    private static final ThreadLocal<EventEmitter> eventEmitters = ThreadLocal.withInitial(EventEmitter::new);
```

Comment:
Hmm.  I'm not feeling particularly safe about this. 
- Are you sure this works properly with reactive pipelines?
- I think we need a mechanism for clearing the value, no? In WildFly, an application can get undeployed. If we don't have a hook in place to clear thread locals during undeploy, this will result in a memory leak. But right, we haven't been putting too much effort in supporting multi-deployment scenarios with WildFly, so maybe it's not that much of a problem.

It's certainly not safe for multi-deployment scenarios, but if it at least has a good performance and doesn't break WildFly completely, then we might as well just go with it.

Reply:
I just ran a test inside my Quarkus application, added in all the events, and successfully listened to them. From looking at the event code, it's pretty stateless, so it looks like it's fine to reuse the object for multiple event calls.

The MetricsServices, on the other hand are stateful, and as I fixed in they must have the "end" called, otherwise memory will leak as metric points would just be left open otherwise. The other occurrence of this I know is inside Quarkus [here](https://github.com/quarkusio/quarkus/blob/main/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/AbstractAsyncDataFetcher.java#L51). I can open a PR to close on an exception here if you wish?

I have assumed here that it is fine to track exceptions against the timings of data points. I can happily change it to just remove the data point so exceptions don't count towards timings.

 Regarding removing from the ThreadLocal, assuming that Wildfly removes the associated threads from an application when it unloaded, all the values should be removed from ThreadLocal automatically. This is what is mentioned in the docs:
 > after a thread goes away, all of its copies of thread-local instances are subject to garbage collection (unless other references to these copies exist).

Reply:
Ok let's not dive too much into WildFly potential memory leaks, most SmallRye projects don't even claim they support any kind of multi-deployment scenario, including undeploying and redeploying the same app, because we are focused on microservices, not traditional application servers. 

Yeah I think we count "exceptional" executions towards the metrics normally just like the successful ones. If there's some cases where a metric measurement isn't ended, that should be fixed

Reply:
Awesome! I'll assemble a patch for Quarkus to ensure that the metrics are always ended!

Reply:
Have created a fix here: https://github.com/quarkusio/quarkus/pull/38686